### PR TITLE
Fix TagFinder.findTagAroundPosition for better emoji support

### PR DESCRIPTION
**Pull request by BazinC**
Fixes https://github.com/superlistapp/super_editor/issues/1863 and https://github.com/superlistapp/super_editor/issues/1882

- fixed findTagAroundPosition to support of emojis
- Widget Text: added "emojis" group and "with consecutive triggers" in "commits" group.
There was edge case in findTagAroundPosition looping through the parameter text code point one by one.
Fixed it with characters iterators.

### CODE REVIEW

**Comment by BazinC** [comment]
golden test fails, I didn't manage to update them.. Can you help me with it?

**Comment by BazinC** [comment]
@matthew-carroll  @angelosilvestre

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart [hunk 1]
```diff
@@ -1042,19 +1105,222 @@ void main() {
...
+        expect(commitedTags.length, 1);
+
+        final commitTag = commitedTags.first;
+
+        // Ensure the committed tag is the emoji and the composing tag is removed
+        expect(commitTag, const IndexedTag(Tag('@', '💙'), '1', 0));
+        expect(plugin.tagIndex.composingStableTag.value, isNull);
+      });
+
+      testWidgetsOnAllPlatforms("before trigger", (tester) async {
```
Comment (matthew-carroll) [inline super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart:1217]: I think my comments above apply to the remaining tests, too, so instead of repeating them I'll just ask that you go through all of this test code and update accordingly.

File: super_editor/lib/src/default_editor/text_tokenizing/tags.dart [hunk 2]
```diff
@@ -23,55 +22,51 @@ class TagFinder {
...
+      }
 
-        if (upstreamCharacter == tagRule.trigger) {
-          // The character we just added to the token bounds is the trigger.
-          // We don't want to move the start any further upstream.
-          break;
-        }
+      if (currentCharacter == tagRule.trigger) {
+        // The character we are reading is the trigger.
+        // We move the iteratorUpstream one last time to include the trigger in the tokenRange and stop looking any further upstream
```
Comment (matthew-carroll) [inline super_editor/lib/src/default_editor/text_tokenizing/tags.dart:54]: This still confuses me. You just checked the `currentCharacter`, so isn't the trigger already in the tokenRange?

File: super_editor/lib/src/default_editor/text_tokenizing/tags.dart [hunk 3]
```diff
@@ -23,55 +22,51 @@ class TagFinder {
...
     }
 
-    while (tokenEndOffset < rawText.length - 1) {
-      final downstreamCharacterIndex = rawText.moveOffsetDownstreamByCharacter(tokenEndOffset)!;
-      final downstreamCharacter = rawText[downstreamCharacterIndex];
-      if (downstreamCharacter != tagRule.trigger && tagRule.excludedCharacters.contains(downstreamCharacter)) {
+    // Move downstream the caret position until we find excluded character or reach the end of the text.
+    while (iteratorDownstream.moveNext()) {
+      final current = iteratorDownstream.current;
+      if (current != tagRule.trigger && tagRule.excludedCharacters.contains(current)) {
```
Comment (matthew-carroll) [inline super_editor/lib/src/default_editor/text_tokenizing/tags.dart:63]: Why is this condition letting us include more trigger characters?

**Review by matthew-carroll — COMMENTED** [review]
Thanks for moving this PR to `main` - I've done another review.

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart [hunk 4]
```diff
@@ -29,6 +29,25 @@ void main() {
         );
       });
 
+      testWidgetsOnAllPlatforms("can start with a single character", (tester) async {
```
[the remark this answers is not in the document]
Comment (BazinC) [reply super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart:32]: while working on the fix, I arrived to a situation where every test passed but I noticed no compositing tag were detected when typing a single character after the trigger. I noticed and fixed it thanks to your previous review

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart [hunk 5]
```diff
@@ -577,6 +598,48 @@ void main() {
           const SpanRange(7, 11),
         );
       });
+
+      testWidgetsOnAllPlatforms("with consecutive triggers", (tester) async {
```
[the remark this answers is not in the document]
Comment (BazinC) [reply super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart:602]: I needed this test because I broke the StableTagIndex behavior while I thought I fixed the main issue. But no test indicaticated something was broken. I added this test to ensure this behavior. I moved it after your comment into 'commit' group test.
I don't see any existing consecutive trigger characters test.

File: super_editor/lib/src/default_editor/text_tokenizing/tags.dart (hunk 2 above)
Comment (BazinC) [reply super_editor/lib/src/default_editor/text_tokenizing/tags.dart:54]: It is because when calculating tokenRange we use` iteratorUpstream.stringAfterLength `to count how much we moved upstream from the `splitIndex`. 
```
Let's say we have this document.
MutableDocument(
  nodes: [
    ParagraphNode(
      id: "1",
      text: AttributedText("Bonjour @Matthew"),
    ),
  ],
);
```
And we place the caret at the end of the paragraph.
```
await tester.placeCaretInParagraph("1", 16);
```

`TagUserReaction` will trigger `_adjustTagAttributionsAroundAlteredTags` -> `_findComposingTagAtCaret` -> `TagFinder.findTagAroundPosition`.

at the beginning of `TagFinder.findTagAroundPosition`, we have:
`splitIndex: 16`
and the `StringCharacterRange` `iteratorUpstream` state is something like:
```
    _end:
    16
    _start:
    16
    _string:
    "Bonjour @Matthew"
    current:
    ""
    stringAfter:
    ""
    stringAfterLength:
    0
    stringBefore:
    "Bonjour @Matthew"
    stringBeforeLength:
    16
```
   
Here `_start` and `_end` are the indexes of range in string, and `_current` represent the substring between `_start` and `_end`.
  
Reading a first character back will move `_start` to 15 et `_end` stays at 16. The current character returned by the iterator is the one between those 2 indexes -> "w".
  
The `while` loop will keep the iterator moving upstream until `current == '@'`.
At this point, iteratorUpstream state is:
```
    _end:
    9
    _start:
    8
    _string:
    "Bonjour @Matthew"
    current:
    "@"
    stringAfter:
    "Matthew"
    stringAfterLength:
    7
    stringBefore:
    "Bonjour "
    stringBeforeLength:
    8
```
    
If we want the '`@`' character to be counted in the iterator `stringAfterLength`, we need to call `moveBack` once more so it moves `_end` index before @.
See stringAfterLength implementation: 
  `int get stringAfterLength => _string.length - _end;`

File: super_editor/lib/src/default_editor/text_tokenizing/tags.dart (hunk 3 above)
Comment (BazinC) [reply super_editor/lib/src/default_editor/text_tokenizing/tags.dart:63]: Good question. I took inspiration from previous of the implementation. I removed it and the tests still pass 👍

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart [hunk 6]
```diff
@@ -100,7 +119,9 @@ void main() {
               ),
             ],
           ),
-          const TagRule(trigger: "@"),
+          plugin: StableTagPlugin(
```
[the remark this answers is not in the document]
Comment (BazinC) [reply super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart:122]: Because I changed _pumpTestEditor signature to take an optional SuperEditorPlugin. It was hard to keep troubleshoot emojis bugs and this way in 'support emoji' test I have access to the plugin and make useful expect checks on the committed tags and composing tag.

**Comment by BazinC** [comment]
Ready for another review @matthew-carroll

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart:32]: Please don't resolve a comment unless you're absolutely sure the issue is resolved.

Based on your answer I have multiple questions.

First, the behavior that's being tested here doesn't seem to be broken for me:

https://github.com/superlistapp/super_editor/assets/7259036/96970fac-ce4e-4c82-8f16-795c5e4adbeb

Second, when you say "@j" wasn't tagged for you, does that also mean that just "@" wasn't tagged for you, either? The composing attribution should be applied as soon as the trigger is typed. You shouldn't need to type any additional character to start composing a tag. So what exactly was working vs not working for you?

If you're trying to ensure that a tag starts composing as early as its supposed to, then you probably want to test "@", not "@j". Moreover, that test would be named "starts when the trigger character is inserted". Lastly, given that such a test verifies the initial behavior for all tag entry, it should be the first test under "composing >".

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart:122]: I don't understand this explanation. Please let me know the specific thing you needed to do that you couldn't do with the previous API. If you only changed this because you were temporarily hacking with some other behaviors, then please revert this to its original API.

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart [hunk 7]
```diff
@@ -1042,19 +1105,222 @@ void main() {
         );
       });
     });
+
+    group("emoji >", () {
+      testWidgetsOnAllPlatforms("typed as first character of a paragraph", (tester) async {
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart]: The updated code adds a few more words but still doesn't provide all the useful information to the reader. You mentioned the editor crashing, but you didn't mention *why* the editor would crash. In theory, we could describe every test as "make sure the editor doesn't crash". The operative detail is why this would crash the editor.

It's good you mentioned the crash in the test name. It's good you linked to the issue ticket.

However, instead of repeating stuff about the editor crashing in the code comment, you should mention what kind of crash you're talking about and/or why it would happen.

Also, in the first test, you failed to mention the relevance of it being the "first character of a paragraph". In the second test, you failed to mention the relevance of moving the caret. Why are these details important? Future readers need to know why these specific details and actions were tested.

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart [hunk 8]
```diff
@@ -1042,19 +1105,259 @@ void main() {
...
+            ),
+          ),
+        );
+
+        // Ensure the paragraph string is well formed: 💙
+        final text = SuperEditorInspector.findTextInComponent("1");
+        expect(text.text, "💙");
+      });
+
+      testWidgetsOnAllPlatforms("can be captured with trigger", (tester) async {
```
Comment (matthew-carroll) [inline super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart:1197]: You added more words to this name, but it doesn't really convey any more information. As a reader of this test, I don't know what "capture" means, and it's not clear why such a test needs to be special for emojis. Please be more specific about what you're verifying in this test.

File: super_editor/lib/src/default_editor/text_tokenizing/tags.dart (hunk 2 above)
Comment (matthew-carroll) [reply super_editor/lib/src/default_editor/text_tokenizing/tags.dart:54]: I tried to read through this but I didn't have the time to carefully check every number in here. I can see that you're trying to make some point about the index vs character, but it's too difficult for me to figure out what you're trying to say.

Can you please try to describe the index vs character issue more succinctly?

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart [hunk 9]
```diff
@@ -1042,19 +1105,222 @@ void main() {
...
+              ParagraphNode(
+                id: "1",
+                text: AttributedText("💙"),
+              ),
+            ],
+          ),
+          plugin: plugin,
+        );
+
+        // Place the caret after the emoji: 💙|
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart:1166]: I unresolved this comment. I don't see any obvious change to address what I said above.

**Comment by matthew-carroll** [comment]
@BazinC any progress on this?

**Comment by matthew-carroll** [comment]
@BazinC checking in one last time - do you plan to update this?

**Comment by BazinC** [comment]
> @BazinC checking in one last time - do you plan to update this?

Sorry @matthew-carroll I didn't have time yet. I will give another try tomorrow (I'm afk right now)

File: super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (BazinC) [reply super_editor/test/super_editor/text_entry/tagging/stable_tags_test.dart:602]: I use consecutives triggers in our app to support @user/@@task/@@@doc. It's a hack.
I understand your point, I'm removing this test.
We should then think of fixing this behavior for @ in the middle of a sentence without triggering a tag.

**Comment by matthew-carroll** [comment]
@BazinC please be sure to re-request a review whenever you think you've handled the feedback.

**Comment by matthew-carroll** [comment]
I'm gonna take over this PR to get it merged in: https://github.com/superlistapp/super_editor/pull/2046
